File: src/main/java/com/protryon/jasm/AType.java
package com.protryon.jasm;

public final class AType {

    public final JType type;
    public final int index;

    private AType(JType type, int index) {
        this.type = type;
        this.index = index;
    }

    public static final AType BOOLEAN = new AType(JType.booleanT, 4);
    public static final AType CHAR = new AType(JType.charT, 5);
    public static final AType FLOAT = new AType(JType.floatT, 6);
    public static final AType DOUBLE = new AType(JType.doubleT, 7);
    public static final AType BYTE = new AType(JType.byteT, 8);
    public static final AType SHORT = new AType(JType.shortT, 9);
    public static final AType INT = new AType(JType.intT, 10);
    public static final AType LONG = new AType(JType.longT, 11);

    public static AType from(int x) {
        switch (x) {
            case 4:
                return BOOLEAN;
            case 5:
                return CHAR;
            case 6:
                return FLOAT;
            case 7:
                return DOUBLE;
            case 8:
                return BYTE;
            case 9:
                return SHORT;
            case 10:
                return INT;
            case 11:
                return LONG;
            default:
                throw new UnsupportedOperationException("invalid atype: " + x);
        }
    }

}

File: src/main/java/com/protryon/jasm/AType.kt
package com.protryon.jasm

class AType private constructor(val type: JType, val index: Int) {
    companion object {

        val BOOLEAN = AType(JType.booleanT, 4)
        val CHAR = AType(JType.charT, 5)
        val FLOAT = AType(JType.floatT, 6)
        val DOUBLE = AType(JType.doubleT, 7)
        val BYTE = AType(JType.byteT, 8)
        val SHORT = AType(JType.shortT, 9)
        val INT = AType(JType.intT, 10)
        val LONG = AType(JType.longT, 11)

        fun from(x: Int): AType {
            when (x) {
                4 -> return BOOLEAN
                5 -> return CHAR
                6 -> return FLOAT
                7 -> return DOUBLE
                8 -> return BYTE
                9 -> return SHORT
                10 -> return INT
                11 -> return LONG
                else -> throw UnsupportedOperationException("invalid atype: $x")
            }
        }
    }

}
